cmd/kubeadm/app/phases/controlplane/components.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controlplane
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	kubeletconfig "k8s.io/kubelet/config/v1beta1"
+
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util/staticpod"
+)
+
+// ControlPlaneComponents contains all components in control plane
+var ControlPlaneComponents = []string{
+	kubeadmconstants.KubeAPIServer,
+	kubeadmconstants.KubeControllerManager,
+	kubeadmconstants.KubeScheduler,
+}
+
+type component struct {
+	name    string
+	labels  map[string]string
+	touched bool
+}
+
+// WaitForControlPlaneComponents wait for control plane component to be ready by check pod status returned by kubelet
+func WaitForControlPlaneComponents(componentNames []string, timeout time.Duration, manifestDir, kubeletDir, certificatesDir string) error {
+	certFile := filepath.Join(certificatesDir, kubeadmconstants.APIServerKubeletClientCertName)
+	keyFile := filepath.Join(certificatesDir, kubeadmconstants.APIServerKubeletClientKeyName)
+
+	client, err := rest.HTTPClientFor(&rest.Config{
+		TLSClientConfig: rest.TLSClientConfig{
+			CertFile: certFile,
+			KeyFile:  keyFile,
+			Insecure: true,
+		},
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to create kubelet client")
+	}
+
+	kubeletEndpoint, err := getKubeletEndpoint(filepath.Join(kubeletDir, kubeadmconstants.KubeletConfigurationFileName))
+	if err != nil {
+		return errors.Wrap(err, "failed to get kubelet endpoint")
+	}
+
+	components := make([]*component, len(componentNames))
+	for i, name := range componentNames {
+		labels, err := getComponentLabels(name, manifestDir)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get pod labels of %s component", name)
+		}
+
+		components[i] = &component{name, labels, false}
+	}
+
+	return wait.PollUntilContextTimeout(context.Background(), 5*time.Second, timeout, false, func(ctx context.Context) (bool, error) {
+		klog.V(1).Infoln("[control-plane] polling status of control plane components...")
+
+		resp, err := client.Get(kubeletEndpoint)
+		if err != nil {
+			fmt.Printf("[kubelet client] Error getting pods [%v]\n", err)
```

Comment:
instead of printing these errors we can store the last error as a `lastError` and return it?
```
err := wait.PollUntilContextTimeout...

if err != nil {
    return lastError
}
```